### PR TITLE
CLI improvements

### PR DESCRIPTION
**Pull request by MarinPostma**
I've made the following changes to the cli:
- `settings-update` become `settings`, with two subcommands: `update` and `show`.
- `document-addition` becomes `documents` with a subcommands: `add` (I'll add a feature to list documents later)
- `search` now has an interactive mode `-i`
- search return the number of documents and the time it took to perform the search.

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -32,9 +36,108 @@ struct Cli {
 
 #[derive(Debug, StructOpt)]
 enum Command {
-    DocumentAddition(DocumentAddition),
+    Documents {
+        #[structopt(subcommand)]
+        cmd: Documents,
+    },
```
Comment (Kerollmops) [inline cli/src/main.rs:42]:
```suggestion
    #[structopt(subcommand)]
    Documents(Documents),
```

File: cli/src/main.rs [hunk 2]
```diff
@@ -32,9 +36,108 @@ struct Cli {
...
     Search(Search),
-    SettingsUpdate(SettingsUpdate),
+    Settings {
+        #[structopt(subcommand)]
+        cmd: Settings,
+    },
```
Comment (Kerollmops) [inline cli/src/main.rs:47]:
```suggestion
    #[structopt(subcommand)]
    Settings(Settings),
```

File: cli/src/main.rs [hunk 3]
```diff
@@ -32,9 +36,108 @@ struct Cli {
...
+}
+
+impl Performer for Command {
+    fn perform(self, index: Index) -> Result<()> {
+        match self {
+            Command::Documents { cmd } => cmd.perform(index),
+            Command::Search(cmd) => cmd.perform(index),
+            Command::Settings { cmd } => cmd.perform(index),
```
Comment (Kerollmops) [inline cli/src/main.rs:55]:
```suggestion
            Command::Documents(cmd) => cmd.perform(index),
            Command::Search(cmd) => cmd.perform(index),
            Command::Settings(cmd) => cmd.perform(index),
```

File: cli/src/main.rs [hunk 4]
```diff
@@ -247,29 +346,83 @@ struct Search {
     offset: Option<usize>,
     #[structopt(short, long)]
     limit: Option<usize>,
+    #[structopt(short, long, conflicts_with_all=&["query", "filter", "offset", "limit"])]
```
Comment (Kerollmops) [inline cli/src/main.rs:349]:
```suggestion
    #[structopt(short, long, conflicts_with = "query")]
```

Why would it be invalid to specify filters, offsets, or limits?

File: cli/src/main.rs [hunk 5]
```diff
@@ -247,29 +346,83 @@ struct Search {
...
+            let stdin = std::io::stdin();
+            let mut lines = stdin.lock().lines();
+            loop {
+                eprint!("> ");
+                std::io::stdout().flush()?;
+                match lines.next() {
+                    Some(Ok(line)) => {
+                        let now = Instant::now();
+                        let jsons =
+                            Self::perform_single_search(&index, &Some(line), &None, &None, &None)?;
```
Comment (Kerollmops) [inline cli/src/main.rs:365]:
```suggestion
                            Self::perform_single_search(&index, &Some(line), &self.filter, &self.offset, &self.limit)?;
```

**Review by Kerollmops — CHANGES_REQUESTED** [review]

File: cli/src/main.rs (hunk 4 above)
Comment (MarinPostma) [reply cli/src/main.rs:349]: Because it is not yet implemented

Comment (MarinPostma) [reply cli/src/main.rs:349]: My goal was to enable it from the interactive prompt instead, so those will always be ignored

File: cli/src/main.rs (hunk 1 above)
Comment (MarinPostma) [reply cli/src/main.rs]: this is not permitted by structopt

**Review by Kerollmops — APPROVED** [review]
Looks good to me!
bors merge

**Pull request merged**